libs/ts-rest/nest/src/lib/api.decorator.ts

```diff
@@ -115,18 +116,38 @@ const getMethodDecorator = (appRoute: AppRoute) => {
 
 @Injectable()
 export class ApiRouteInterceptor implements NestInterceptor {
+  constructor(private readonly appRoute: AppRoute) {}
+
+  private isAppRouteResponse(
+    value: unknown
+  ): value is { status: HTTPStatusCode; body?: any } {
+    return (
+      value !== null &&
+      typeof value === 'object' &&
+      'status' in value &&
+      typeof value.status === 'number'
+    );
+  }
+
   intercept(context: ExecutionContext, next: CallHandler): Observable<any> {
     const res: Response = context.switchToHttp().getResponse();
 
     return next.handle().pipe(
       map((value) => {
-        if (
-          typeof value === 'object' &&
-          typeof value.status === 'number' &&
-          value.body !== undefined
-        ) {
-          res.status(value.status);
-          return value.body;
+        if (this.isAppRouteResponse(value)) {
+          const statusNumber = value.status;
+
+          const responseValidation = checkZodSchema(
+            value.body,
+            this.appRoute.responses[statusNumber]
+          );
+
+          if (!responseValidation.success) {
+            throw new BadRequestException(responseValidation.error);
```

Comment:
I'd imagine a response validation error would qualify as a 5xx error than than a 400 error

Reply:
Can we make that response error configurable? I can see cases where you'd want to customize the error based on the zod validation.

Reply:
You  could do that already using [exception filters](https://docs.nestjs.com/exception-filters#exception-filters-1) on HttpExceptions and check if `res.body.name === 'ZodError'`, then rewrite the response status and body if needed, but it would be cleaner if we use a custom error type and filter, but I guess that's a discussion for another thread.